Controllers films users

DmitryKarpushov opened the pull request:
Спринт 9

Commit:
1) Сделал контроллер UserController в котором добавил
   -Создание пользователя
   -Отображение всех пользователей
   -Обновление пользователя.
Сделал контроллер FilmController в котором добавил
   -Создание фильма
   -Отображение всех фильмов
   -Обновление фильма.
2)Добавил 2 модели User и Film.
3)Добавил 2 класса тестов UserTest и FilmTest.
4)Добавил DateValidatorUsingDate для валидации выпуска фильма не раньше 1895-12-28.
  Добавил LoginValidatorUsingLogin для валидации логина,который не содержит пробелы.

Commit:
1) Сделал контроллер UserController в котором добавил
   -Создание пользователя
   -Отображение всех пользователей
   -Обновление пользователя.
Сделал контроллер FilmController в котором добавил
   -Создание фильма
   -Отображение всех фильмов
   -Обновление фильма.
2)Добавил 2 модели User и Film.
3)Добавил 2 класса тестов UserTest и FilmTest.
4)Добавил DateValidatorUsingDate для валидации выпуска фильма не раньше 1895-12-28.
  Добавил LoginValidatorUsingLogin для валидации логина,который не содержит пробелы.

Commit:
1) Сделал контроллер UserController в котором добавил
   -Создание пользователя
   -Отображение всех пользователей
   -Обновление пользователя.
Сделал контроллер FilmController в котором добавил
   -Создание фильма
   -Отображение всех фильмов
   -Обновление фильма.
2)Добавил 2 модели User и Film.
3)Добавил 2 класса тестов UserTest и FilmTest.
4)Добавил DateValidatorUsingDate для валидации выпуска фильма не раньше 1895-12-28.
  Добавил LoginValidatorUsingLogin для валидации логина,который не содержит пробелы.

FanaticFiz commented on pom.xml:
```diff
@@ -0,0 +1,67 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Давай поговорим о гите!

Не пиши столь подробные комментарии.
![image](https://user-images.githubusercontent.com/3994663/201535524-291b4d08-1278-4ff0-ac73-2f36202390f1.png)

Тут достаточно описывать кратко.

FanaticFiz replied:
Нормально представься: 
![image](https://user-images.githubusercontent.com/3994663/201535626-d9447cc8-ace0-46d6-a823-c8ba636dca35.png)

Имя/фамилия

FanaticFiz commented on pom.xml:
```diff
@@ -0,0 +1,67 @@
+	<version>0.0.1-SNAPSHOT</version>
+	<name>filmorate</name>
+	<description>Spring Boot filmorate</description>
+	<properties>
```
причеши немного pom файл: очень не хватает отделения друг от друга (пустыми строками) глобальных частей таких как `parent`, `properties`, `dependencies`

FanaticFiz commented on pom.xml:
```diff
@@ -0,0 +1,67 @@
+	</parent>
+	<groupId>ru.yandex.practicum</groupId>
+	<artifactId>filmorate</artifactId>
+	<version>0.0.1-SNAPSHOT</version>
```
Советую прочитать про семантическое версионирование, ну а после не стесняйся - смело проставь версию!

FanaticFiz commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,53 @@
+import org.slf4j.LoggerFactory;
+import org.springframework.web.bind.annotation.*;
+import ru.yandex.practicum.filmorate.datavalidation.ValidationFieldsFilm;
+import ru.yandex.practicum.filmorate.datavalidation.ValidationFieldsUser;
```
Импорты не подчищены...

FanaticFiz commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,53 @@
+    @PostMapping
+    public Film createFilm(@Valid @RequestBody Film film) {
+        logger.info("FilmController.createFilm: Начали добавление фильма");
+        ValidationFieldsFilm.validateFields(film);
```
Если ты собираешься бросать ексепшн, то лучше так и называть метод throwIfNotValidReleaseDate, например

FanaticFiz commented on src/main/java/ru/yandex/practicum/filmorate/datavalidation/ValidationFieldsFilm.java:
```diff
@@ -0,0 +1,36 @@
+ * @author Дмитрий Карпушов 12.11.2022
+ */
+public class ValidationFieldsFilm {
+    private static final Logger logger = LoggerFactory.getLogger(FilmController.class);
```
Почему тут всё "слиплось"?

FanaticFiz commented on src/main/java/ru/yandex/practicum/filmorate/datavalidation/ValidationFieldsFilm.java:
```diff
@@ -0,0 +1,36 @@
+    public static void noFoundFilm(Film film, Map<Integer,Film> films) {
+        if (!films.containsKey(film.getId())) {
+            logger.info("ValidationFieldsFilm.isValidDate: Проверяем на существование фильма");
+            throw new IdValidationException("Такого фильма не существует");
```
Если фильм не найден id не виноват! ))

Эксепшн лучше назвать `NotFoundException` бросать его будем тогда когда не найдена любая из сущностей в приложении...

FanaticFiz commented on src/main/java/ru/yandex/practicum/filmorate/datavalidation/ValidationFieldsFilm.java:
```diff
@@ -0,0 +1,36 @@
+    public static void validateFields(Film film) {
+        logger.info("ValidationFieldsFilm.validateFields: Проверяем на валидность дату");
+        if (isValidDate(film.getReleaseDate())){
+            throw new DateValidationException("Дата релиза не должна быть раньше 28 декабря 1895 года");
```
И тут надо более обще: `BadRequestException`

FanaticFiz commented on src/main/java/ru/yandex/practicum/filmorate/datavalidation/ValidationFieldsUser.java:
```diff
@@ -0,0 +1,39 @@
+            logger.info("ValidationFieldsUser.validateFields: Проверяем на существование пользователя");
+            throw new LoginValidationException("Логин содержит пробелы");
+        }
+        if (user.getName() == null || user.getName().isEmpty() || user.getName().isBlank()){
```
Вот это условие это не валидация, а часть бизнес логики... поэтому уместнее в сервисе.

FanaticFiz commented on src/main/java/ru/yandex/practicum/filmorate/exception/LoginValidationException.java:
```diff
@@ -0,0 +1,10 @@
+/**
+ * @author Дмитрий Карпушов 12.11.2022
+ */
+public class LoginValidationException extends RuntimeException{
```
Этот ексепшн может быть удалён. Будет один общий BadRequestException.
Когда не верно было к нам заслано любое из полей мы его кидаем.

FanaticFiz commented on src/main/resources/application.properties:
```diff
@@ -0,0 +1 @@
+
```
Я буду топить за то чтобы вы сразу привыкали к написанию файла свойств для приложения в формате `yaml -> application.yml`

Соответственно - попрошу тебя этому следовать.Как мини задание: разберись как задать порт для сервиса отличный от дефолтного, и как задать чтобы порт был случайным.(Менять порт на другой не нужно, для тестов нужен именно 8080)

FanaticFiz reviewed (changes requested):
Финальный комментарий на Я.Практикуме

Commit:
1) Сделал контроллер UserController в котором добавил
   -Создание пользователя
   -Отображение всех пользователей
   -Обновление пользователя.
Сделал контроллер FilmController в котором добавил
   -Создание фильма
   -Отображение всех фильмов
   -Обновление фильма.
2)Добавил 2 модели User и Film.
3)Добавил 2 класса тестов UserTest и FilmTest.
4)Добавил DateValidatorUsingDate для валидации выпуска фильма не раньше 1895-12-28.
  Добавил LoginValidatorUsingLogin для валидации логина,который не содержит пробелы.
Обновление:
   Удалил:DateValidatorUsingDate Заменил на: FilmValidator.
   Удалил:LoginValidatorUsingLogin Заменил на: @Pattern в классе User.

FanaticFiz commented on src/main/resources/application.properties:
```diff
@@ -0,0 +1,5 @@
+#?????? ???? ???????? ?? ??????????
+#server.port=8080
```
Это ты так задаешь вопрос?

Надо переименовать файл, как я показывал, и тогда:

```
server:
  port: 8080
```

FanaticFiz commented on src/test/java/ru/yandex/practicum/filmorate/films/FilmTest.java:
```diff
@@ -0,0 +1,73 @@
+import java.util.Set;
+
+/**
+ * @author ������� �������� 13.11.2022
```
???

FanaticFiz commented on src/test/java/ru/yandex/practicum/filmorate/films/FilmTest.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    @Test
+    @DisplayName("5) �������� ���������. �������� ���� �� 1895-12-28")
```
Ты где это набирал?
В других местах нормально же...

Commit:
1) Сделал контроллер UserController в котором добавил
   -Создание пользователя
   -Отображение всех пользователей
   -Обновление пользователя.
Сделал контроллер FilmController в котором добавил
   -Создание фильма
   -Отображение всех фильмов
   -Обновление фильма.
2)Добавил 2 модели User и Film.
3)Добавил 2 класса тестов UserTest и FilmTest.
4)Добавил DateValidatorUsingDate для валидации выпуска фильма не раньше 1895-12-28.
  Добавил LoginValidatorUsingLogin для валидации логина,который не содержит пробелы.
Обновление:
   Удалил:DateValidatorUsingDate Заменил на: FilmValidator.
   Удалил:LoginValidatorUsingLogin Заменил на: @Pattern в классе User.

Commit:
1) Сделал контроллер UserController в котором добавил
   -Создание пользователя
   -Отображение всех пользователей
   -Обновление пользователя.
Сделал контроллер FilmController в котором добавил
   -Создание фильма
   -Отображение всех фильмов
   -Обновление фильма.
2)Добавил 2 модели User и Film.
3)Добавил 2 класса тестов UserTest и FilmTest.
4)Добавил DateValidatorUsingDate для валидации выпуска фильма не раньше 1895-12-28.
  Добавил LoginValidatorUsingLogin для валидации логина,который не содержит пробелы.
Обновление:
   Удалил:DateValidatorUsingDate Заменил на: FilmValidator.
   Удалил:LoginValidatorUsingLogin Заменил на: @Pattern в классе User.
